pkg/helpers/helpers.go

```diff
@@ -0,0 +1,641 @@
+}
+
+// GetMaxConcurrentReconciles get the max concurrent reconciles from MAX_CONCURRENT_RECONCILES env,
+// if the reconciles cannot be found, return 1
+func GetMaxConcurrentReconciles() int {
+	maxConcurrentReconciles := 1
+	if os.Getenv(maxConcurrentReconcilesEnvVarName) != "" {
+		var err error
+		maxConcurrentReconciles, err = strconv.Atoi(os.Getenv(maxConcurrentReconcilesEnvVarName))
+		if err != nil {
+			klog.Warningf("The value of %s env is wrong, using default reconciles (1)", maxConcurrentReconcilesEnvVarName)
+			maxConcurrentReconciles = 1
+		}
+	}
+	return maxConcurrentReconciles
+}
+
+// SetLabels set a label on the given object
+func SetLabels(objMeta *metav1.ObjectMeta, key, val string) bool {
+	lables := objMeta.GetLabels()
+	if lables == nil {
+		lables = make(map[string]string)
+	}
+
+	currentVal, ok := lables[key]
+	if !ok || currentVal != val {
+		lables[key] = val
+		objMeta.SetLabels(lables)
+		return true
+	}
+
+	return false
+}
+
+// SetAnnotations set an annotation on the given object
+func SetAnnotations(objMeta *metav1.ObjectMeta, key, val string) bool {
+	annotations := objMeta.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	currentVal, ok := annotations[key]
+	if !ok || currentVal != val {
+		annotations[key] = val
+		objMeta.SetAnnotations(annotations)
+		return true
+	}
+
+	return false
+}
+
+// GenerateClientFromSecret generate a client from a given secret
+func GenerateClientFromSecret(secret *corev1.Secret) (client.Client, meta.RESTMapper, error) {
+	var err error
+	var config *clientcmdapi.Config
+
+	if kubeconfig, ok := secret.Data["kubeconfig"]; ok {
+		config, err = clientcmd.Load(kubeconfig)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	token, tok := secret.Data["token"]
+	server, sok := secret.Data["server"]
+	if tok && sok {
+		config = clientcmdapi.NewConfig()
+		config.Clusters["default"] = &clientcmdapi.Cluster{
+			Server:                string(server),
+			InsecureSkipTLSVerify: true,
+		}
+		config.AuthInfos["default"] = &clientcmdapi.AuthInfo{
+			Token: string(token),
+		}
+		config.Contexts["default"] = &clientcmdapi.Context{
+			Cluster:  "default",
+			AuthInfo: "default",
+		}
+		config.CurrentContext = "default"
+	}
+
+	if config == nil {
+		return nil, nil, fmt.Errorf("kubeconfig or token and server are missing")
+	}
+
+	clientConfig, err := clientcmd.NewDefaultClientConfig(*config, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	runtimeClient, err := client.New(clientConfig, client.Options{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(clientConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return runtimeClient, mapper, nil
+}
+
+// AddManagedClusterFinalizer add a finalizer to a managed cluster
+func AddManagedClusterFinalizer(client client.Client, recorder record.EventRecorder,
```

Comment:
if it is using controller-runtime, there is a util func here  https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L352

Reply:
will use them

Reply:
do you need an event when adding a finalizer?

Reply:
I hope to record an event when I add/remove a finalizer from managed cluster